PROPOSAL -- airgap lite lab with ability for extra releases (#53)
replicatedhq/kots-field-labs: labs/lab08-airgap-lite/README.md

You will have received the IP of a jump box and the name of an Air Gap server.
For example, you may have received:

```text
dx411-dex-lab05-airgap-jump = 104.155.131.205
dx411-dex-lab05-airgap
```

In general, the name of the private server will be the same as the jump box, with the characters `-jump` removed from the suffix.
Put another way, you could construct both instance names programatically as

```shell
${REPLICATED_APP}-lab05-airgap-jump
```

and

```shell
${REPLICATED_APP}-lab05-airgap
```

### Connecting

First set your application slug, and the public IP of your jump box:

```shell
export JUMP_BOX_IP=...
export REPLICATED_APP=... # your app slug
```

Next, you can SSH into the Air Gap server using the following command:

```shell
ssh -J kots@${JUMP_BOX_IP} kots@${REPLICATED_APP}-lab05-airgap
```
> **NOTE**: You will be prompted **twice** to change the password on first login. Once for the JUMP_BOX and again for the AIRGAP server.

Once you're on the Air Gap server, you can verify that the server indeed does not have internet access. Once you're convinced, you 
can ctrl+C the command and proceed to the next section

```shell
curl -v https://kubernetes.io

> ```

Reviewer: Replace with

First set your application slug, the public IP of your jump box and your first name:

```shell
export JUMP_BOX_IP=...
export REPLICATED_APP=... # your app slug
export FIRST_NAME=... # your first name
```

Next, you can SSH into the Air Gap server using the following command:

```shell
ssh -J ${FIRST_NAME}@${JUMP_BOX_IP} ${FIRST_NAME}@${REPLICATED_APP}-lab08-airgap-lite
```

The `-J` option, allows to connect to the target host by first making a ssh connection to the jump host (`${JUMP_BOX_IP}`) described by destination and then establishing a TCP forwarding to the ultimate destination (`${REPLICATED_APP}-lab08-airgap-lite`) from there.

You can also do it in multiple steps and achieve the same:

```shell
local> ssh ${FIRST_NAME}@${JUMP_BOX_IP}
jump> export REPLICATED_APP=...
jump> ssh ${REPLICATED_APP}-lab08-airgap-lite
```

Once you're on the Air Gap server, you can verify that the server indeed does not have internet access. Once you're convinced, you 
can ctrl+C the command and proceed to the next section

```shell
curl -v https://kubernetes.io
```

***
## Moving Assets into place

If you haven't already, you can log out of the Air Gap instance with `exit` or ctrl+D. 
Our next step is to collect the assets we need for an Air Gap installation:

1. A license with the Air Gap entitlement enabled
2. An Air Gap bundle containing the kURL cluster components
3. An Air Gap bundle containing the application components

(2) and (3) are separate artifacts to cut down on bundle size during upgrade scenarios where only the application version 
is changing and no changes are needed to the underlying cluster.

#### Starting the kURL Bundle Download

Now, let's SSH to our jump box (the one with the public IP) `ssh kots@<jump box IP address>` and download the kurl bundle.
Replace the URL below with the one you can query from 

```
replicated channel inspect lab05-airgap
```

```text

> kots@dx411-dex-lab05-airgap-jump ~$ curl -o kurlbundle.tar.gz <URL>

Reviewer: ```suggestion
kots@dx411-dex-lab08-airgap-lite-jump ~$ curl -o kurlbundle.tar.gz <URL>
```

```

This will take several minutes, leave this running and proceed to the next step, we'll come back in a few minutes.

#### Building an Airgap Release

By default, only the Stable and Beta channels will automatically build Air Gap bundles

- manually build
- set channel to auto build

For a production application, Air Gap releases will be built automatically on the Stable channel, so this won't
be necessary.

> In this case, since we're working off the `lab05-airgap` channel, you'll want to enable Air Gap builds on that channel.

Reviewer: ```suggestion
In this case, since we're working off the `lab08-airgap-lite` channel, you'll want to enable Air Gap builds on that channel.
```

You can check the build status by navigating to the "Release History" for the channel.

![release-history](img/channel-release-history.png)

You can build invividual bundles on the Release History page, but you'll likely want to edit the channel and enable "build all releases for this channel".

![edit-channel](img/channel-edit-info-btn.png)

![auto-build](img/channel-enable-airgap.png)

Now you should see all the bundles building or built on the release history page:

![airgap-built](img/airgap-builds.png)

#### Enabling Airgap for a customer

> The first step will be to enable Air Gap for the `lab5` customer:

Reviewer: ```suggestion
The first step will be to enable Air Gap for the `lab8` customer:
```

![enable-airgap](./img/airgap-customer-enable.png)

#### Download Airgap Assets 
After saving the customer, scroll to the bottom of the page to the `Download Portal` section.

![download-portal](img/airgap-customer-portal.png)

Generate a new password and save it somewhere in your notes.
Next, click the link to open the download portal. 
This is a link you would usually send to your customer, so from here on we'll be wearing our "end user" hat.

Navigate to the "embedded cluster" option and review the three downloadable assets.

![download-portal-view](img/download-portal-view.png)

Download the license file, but **don't download the kURL bundle** -- this is the download we already started on the server.

> You'll also want to download the other bundle `Latest Lab 1.5: Airgap Bundle` to your workstation.

Reviewer: ```suggestion
You'll also want to download the other bundle `Latest Lab 1.8: Airgap Bundle` to your workstation.
```

Now, let's SSH to our jump box (the one with the public IP) `ssh kots@<jump box IP address>` and download the kurl bundle.
Replace the URL with the one you copied above.

At the beginning of the lab, we downloaded the bundle with this command from the Jump box.

```text
kots@dx411-dex-lab05-airgap-jump ~$ curl -o kurlbundle.tar.gz <URL>
```

It should be finished now, so you can copy it to the Air Gap server. 
You can use the DNS name in this case, as described in [Instance Overview](#instance-overview).

```text

> kots@dx411-dex-lab05-airgap-jump ~$ scp kurlbundle.tar.gz kots@dx411-dex-lab05-airgap:/home/kots

Reviewer: ```suggestion
kots@dx411-dex-lab08-airgap-lite-jump ~$ scp kurlbundle.tar.gz kots@dx411-dex-lab08-airgap-lite:/home/kots
```

```

> **Note**: -- we use SCP via an SSH tunnel in this case, but the Air Gap methods in this lab also extend to
more locked down environments where e.g. physical media is required to move assets into the datacenter.

Now we'll SSH all the way to Air Gap node. If you still have a shell on your jump box, you can use the instance name.

```text

> kots@dx411-dex-lab05-airgap-jump ~$ ssh dx411-dex-lab05-airgap

Reviewer: ```suggestion
kots@dx411-dex-lab08-airgap-lite-jump ~$ ssh dx411-dex-lab08-airgap-lite
```

```

Otherwise, you can use the one below 

```shell

> ssh -J kots@lab05-airgap-jump kots@${REPLICATED_APP}-lab05-airgap

Reviewer: ```suggestion
ssh -J ${FIRST_NAME}@lab08-airgap-lite-jump ${FIRST_NAME}@${REPLICATED_APP}-lab08-airgap-lite
```

```

Once you're on the Air Gap node, untar the bundle and run the install script with the `airgap` flag.
kURL install flags are documented [in the kurl.sh docs](https://kurl.sh/docs/install-with-kurl/advanced-options).

```shell
tar xvf kurlbundle.tar.gz
sudo bash install.sh airgap
```

At the end, you should see a `Installation Complete` message as shown below. Since the instance is Air Gap, we'll need to create a port forward to access the UI from your workstation in the next step.

![kurl-password](img/kurl-password.png)

***
## Accessing the UI via SSH tunnel, Configuring the instance

You'll want to create a port forward from your workstation in order to access to UI locally.
Again we'll use `REPLICATED_APP` to construct the DNS name but you can input it manually as well.

```shell

> export JUMP_BOX_IP=... # your jumpbox IP
> export REPLICATED_APP=... # your app slug
> ssh -NL 8800:${REPLICATED_APP}-lab05-airgap:8800 -L 8888:${REPLICATED_APP}-lab05-airgap:8888 kots@${JUMP_BOX_IP}

Reviewer: ```suggestion
export FIRST_NAME=... # your firstname (lowercase)
export JUMP_BOX_IP=... # your jumpbox IP
export REPLICATED_APP=... # your app slug
ssh -NL 8800:${REPLICATED_APP}-lab08-airgap-lite:8800 -L 8888:${REPLICATED_APP}-lab08-airgap-lite:8888 ${FIRST_NAME}@${JUMP_BOX_IP}
```

![airgap-upload](img/airgap-upload.png)

You'll see the bundle uploaded and images being pushed to kURL's internal registry. This will take a few minutes to complete.

![airgap-push](img/airgap-push.png)

Once uploaded `Preflight Checks` will run. These are designed to ensure this server has the minimum system and software requirements to run the application.
Depending on your YAML in `preflight.yaml`, you may see some of the example preflight checks fail.
If you have failing checks, you can click continue -- the UI will show a warning that will need to be dismissed before you can continue.

![Preflight Checks](https://kots.io/images/guides/kots/preflight.png)

We'll find that the application is unavailable. 

![app-down](img/app-down.png)

While we'll explore [support techniques for airgapped environments](#collecting-a-cli-support-bundle) 
below, in this case you should observe that our deployment is simply not valid, specifically, the
standard nginx entrypoint has been overriden:

```yaml
      containers:
        - name: nginx
          image: nginx
          command:
            - exit
            - "1"
```

> So we'll need to create a new release in order to fix this.

Reviewer: ```suggestion
So we'll need to deploy a new release in order to fix this.
```

***
## Deploying a new version

From the `labs/lab05-airgap` directory, update the `manifests/nginx-deployment.yaml` file to remove the command override as shown below.

```diff
diff --git a/labs/lab05-airgap/manifests/nginx-deployment.yaml b/labs/lab05-airgap/manifests/nginx-deployment.yaml
index fa29e8d..3a66405 100644
--- a/labs/lab05-airgap/manifests/nginx-deployment.yaml
+++ b/labs/lab05-airgap/manifests/nginx-deployment.yaml
@@ -16,9 +16,6 @@ spec:
       containers:
         - name: nginx
           image: nginx:latest
-          command:
-            - exit
-            - "1"
           volumeMounts:
             - mountPath: /usr/share/nginx/html
               name: html
```

Once you're satisfied with your `nginx-deployment.yaml` create a new release with `make release`.
You'll need to ensure you have your `REPLICATED_APP` and `REPLICATED_API_TOKEN` set. See the **Getting Started** section for information on how to obtain and set these.

```shell
make release

> ```

Reviewer: Replace with:

We already created a new Release in Vendor Portal with the fixed container. In order to make the release available, go to `Releases > Sequence 2` and click `Promote`. Select the `lab08-airgap` channel to promote it to.

Author: @jdewinne wanna pull this in yourself on the next iteration?

> Once the release is made, you should be able to navigate back to the customer download portal we accessed from the customer page.

Reviewer: ```suggestion
Once the promotion is done, you should be able to navigate back to the customer download portal we accessed from the customer page.
```

> Scrolling to the bottom, you can click "show older bundles" to see the history of releases on the lab05-airgap channel.

Reviewer: ```suggestion
Scrolling to the bottom, you can click "show older bundles" to see the history of releases on the lab08-airgap channel.
```

The new release may take a minute or two to build, so you're want to refresh the make until you see one

> with a timestamp that matches when you ran `make release`.

Reviewer: ```suggestion
with a timestamp that matches when you promoted the release.
```

![download-portal-more](img/download-portal-more.png)

Once you've downloaded the new version, in the KOTS Admin Console select **Version History** and click "**Upload a new version**" and select your bundle.

![airgap-new-upload](img/airgap-new-upload.png)

You'll see the bundle upload as before and you'll have the option to deploy it once the
preflight checks complete. Click **Deploy** to perform the upgrade.

Click the **Application** button to navigate back to the main landing page. The app should now show as **Ready** status on the main dashboard.

> In order to access the application select **Open Lab 5**. 

Reviewer: ```suggestion
In order to access the application select **Open Lab 8**. 
```
